src/dyngen.cpp

```diff
@@ -14,199 +14,386 @@ static CodeLibrary *gLibrary = nullptr;
 // NRT callbacks for DynGen init
 // *****************************
 
-// Gets called deferred from the UGen init in a
-// stage2 thread which is NRT - so it is safe to allocate
-// memory and also spawn a thread to which we offload
-// all the heavy lifting of initializing the VM.
-bool dynGenCompileCallback(World *world, void *rawCallbackData) {
+// stage 2 - NRT
+bool createVmAndCompile(World* world, void *rawCallbackData) {
   auto callbackData = static_cast<DynGenCallbackData*>(rawCallbackData);
 
-  CodeLibrary* node = gLibrary;
-  while (node && node->id != callbackData->scriptHash) {
-    node = node->next;
-  }
+  callbackData->vm = new EEL2Adapter(
+    callbackData->numInputChannels,
+    callbackData->numOutputChannels,
+    callbackData->sampleRate,
+    callbackData->blockSize,
+    callbackData->world,
+    callbackData->parent
+  );
 
-  if (!node) {
-    Print("Could not find script with hash %i\n", callbackData->scriptHash);
-    return false;
+  callbackData->vm->init(callbackData->code);
+  // continue to stage 3
+  return true;
+}
+
+// stage 3 - RT
+bool swapVmPointers(World* world, void *rawCallbackData) {
+  auto callbackData = static_cast<DynGenCallbackData*>(rawCallbackData);
+  // only replace if DynGen instance is still existing
+  if (callbackData->dynGenStub->mObject) {
+    callbackData->oldVm = callbackData->dynGenStub->mObject->mVm;
+    callbackData->dynGenStub->mObject->mVm = callbackData->vm;
   }
+  return true;
```

Comment:
If the DynGen object is gone, we have to free the VM we just created. Here's a simple solution:
```
// in swapVmPointers():

if (callbackData->dynGenStub->mObject) {
    callbackData->oldVm = callbackData->dynGenStub->mObject->mVm;
    callbackData->dynGenStub->mObject->mVm = callbackData->vm;
    callbackData->vm = nullptr; // see deleteOldVm()
}

// in deleteOldVm():

// in case we could not move the VM to the DynGen instance in swapVmPointers()
delete callbackData->vm;
```

Reply:
Don't you mean

```diff
diff --git a/src/dyngen.cpp b/src/dyngen.cpp
index b078a7a..4f3388f 100644
--- a/src/dyngen.cpp
+++ b/src/dyngen.cpp
@@ -39,6 +39,10 @@ bool swapVmPointers(World* world, void *rawCallbackData) {
   if (callbackData->dynGenStub->mObject) {
     callbackData->oldVm = callbackData->dynGenStub->mObject->mVm;
     callbackData->dynGenStub->mObject->mVm = callbackData->vm;
+  } else {
+    // mark the vm we just created ready for deletion since the DynGen
+    // it was created for does not exist anymore.
+    callbackData->oldVm = callbackData->vm;
   }
   return true;
 }
```

Reply:
This would leak the old VM. We have to delete both.

Reply:
My proposed solution was a mistake. Your proposed code is correct.